import_export/resources.py

```diff
@@ -252,8 +252,13 @@ def init_instance(self, row=None):
 
     def get_instance(self, instance_loader, row):
         """
-        Calls the :doc:`InstanceLoader <api_instance_loaders>`.
+        If all 'import_id_fields' are present in the dataset, calls
+        the :doc:`InstanceLoader <api_instance_loaders>`. Otherwise,
+        returns `None`.
         """
+        for field_name in self.get_import_id_fields():
+            if field_name not in row:
```

Comment:
Wait isn’t `row` potentially keyed with `column_names` that would only _indirectly_ match `imported_id_fields`, which in turn reference attributes?

Reply:
Hi @dhorlick. I'm afraid I'm not following. Could you perhaps provide an example of where you'd have indirect matches between `column_names` and `import_id_fields`?

Reply:
Sure. With this patch to tweak the test app and add some console output to the library
```diff
diff --git a/import_export/resources.py b/import_export/resources.py
index 5b3a0b3..61e5d1d 100644
--- a/import_export/resources.py
+++ b/import_export/resources.py
@@ -262,9 +262,12 @@ class Resource(metaclass=DeclarativeMetaclass):
         the :doc:`InstanceLoader <api_instance_loaders>`. Otherwise,
         returns `None`.
         """
+        print(f"Resource.get_instance(): starting. row={row}")
         for field_name in self.get_import_id_fields():
             if field_name not in row:
+                print(f"Oh noes! field_name {field_name} not in {row}")
                 return
+        print("Proceeding to get_instance...")
         return instance_loader.get_instance(row)
 
     def get_or_init_instance(self, instance_loader, row):
diff --git a/tests/core/admin.py b/tests/core/admin.py
index 90f007f..af927a7 100644
--- a/tests/core/admin.py
+++ b/tests/core/admin.py
@@ -6,6 +6,8 @@ from import_export.resources import ModelResource
 from .forms import CustomConfirmImportForm, CustomImportForm
 from .models import Author, Book, Category, Child, EBook
 
+from import_export.fields import Field
+
 
 class ChildAdmin(ImportMixin, admin.ModelAdmin):
     pass
@@ -13,8 +15,14 @@ class ChildAdmin(ImportMixin, admin.ModelAdmin):
 
 class BookResource(ModelResource):
     
+    isbn = Field(attribute='isbn', column_name='ISBN')
+    
     class Meta:
         model = Book
+        fields = ( 'author', 'author_email', 'published', 'name', 'published_time', 'price', 'imported', 'added', 'categories', 'isbn')
+        import_id_fields = (
+            "isbn",
+        )
 
     def for_delete(self, row, instance):
         return self.fields['name'].clean(row) == ''
diff --git a/tests/core/migrations/0009_book_isbn.py b/tests/core/migrations/0009_book_isbn.py
new file mode 100644
index 0000000..8cff711
--- /dev/null
+++ b/tests/core/migrations/0009_book_isbn.py
@@ -0,0 +1,19 @@
+# Generated by Django 2.2.7 on 2019-11-22 09:45
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('core', '0008_auto_20190409_0846'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='book',
+            name='isbn',
+            field=models.CharField(default=None, max_length=13, unique=True, verbose_name='ISBN'),
+            preserve_default=False,
+        ),
+    ]
diff --git a/tests/core/models.py b/tests/core/models.py
index 97f5d5c..1998a4c 100644
--- a/tests/core/models.py
+++ b/tests/core/models.py
@@ -44,6 +44,8 @@ class Book(models.Model):
 
     categories = models.ManyToManyField(Category, blank=True)
 
+    isbn = models.CharField('ISBN', max_length=13, null=False, blank=False, unique=True)
+
     def __str__(self):
         return self.name
```

If add a book and then try to upload this CSV file, to update it
```csv
ISBN,author,name,price
765392763,3,Walkaway: A Novel,1
```

Then I get this app output
```
Line number: 1 - UNIQUE constraint failed: core_book.isbn
765392763, 3, Walkaway: A Novel, 1
Traceback (most recent call last):
File "/Users/dhorlick/diy/test/django-import-export-f/django-import-export-git/venv/src/django-import-export/tests/venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
return self.cursor.execute(sql, params)
File "/Users/dhorlick/diy/test/django-import-export-f/django-import-export-git/venv/src/django-import-export/tests/venv/lib/python3.8/site-packages/django/db/backends/sqlite3/base.py", line 383, in execute
return Database.Cursor.execute(self, query, params)
sqlite3.IntegrityError: UNIQUE constraint failed: core_book.isbn
```

And this console output
```
import_data_inner sez: dataset.dict = [OrderedDict([('ISBN', '765392763'), ('author', '3'), ('name', 'Walkaway: A Novel'), ('price', '1')])]
Resource.get_instance(): starting. row=OrderedDict([('ISBN', '765392763'), ('author', '3'), ('name', 'Walkaway: A Novel'), ('price', '1')])
Oh noes! field_name isbn not in OrderedDict([('ISBN', '765392763'), ('author', '3'), ('name', 'Walkaway: A Novel'), ('price', '1')])
```

If I back-out the pull request change locally, then the upload succeeds.

Reply:
Ah, okay. I see now.

So, instead of just looking for columns matching `import_id_fields`, `get_instance()` need to be checking whether the actual fields named by `import_id_fields` have a `column_name` value, and if so, look for a columns matching that. I will submit another PR.